clis/teliod/src/main.rs

```diff
@@ -0,0 +1,195 @@
+
+/// Starts the Teliod daemon which will run actual Telio.
+///
+/// # Arguments
+///
+/// * `config` - Daemon config.
+fn start_daemon(config: TeliodDaemonConfig) -> Result<()> {
+    println!("Starting Teliod daemon...");
+
+    let (api_fd, daemon_fd) = pipe()?;
+
+    match TeliodDaemon::start() {
+        Ok(ProcessType::Api) => {
+            let _ = close(daemon_fd.as_raw_fd());
+
+            // Daemon sends to API process a message that it is ready
+            let mut api_pipe_end = File::from(api_fd);
+            let mut message = String::new();
+            api_pipe_end.read_to_string(&mut message)?;
+
+            if message == DAEMON_START_MSG {
+                println!("Teliod daemon started");
+            } else {
+                print!("Unexpected pipe message: {}", message);
+            }
+
+            // We really can't do anything with wrong pipe message, so we return Ok anyway
+            Ok(())
+        }
+        Ok(ProcessType::Daemon) => {
+            let _ = close(api_fd.as_raw_fd());
+            daemon_event_loop(config, File::from(daemon_fd))
+        }
+        Err(e) => Err(e),
+    }
+}
+
+fn daemon_event_loop(config: TeliodDaemonConfig, mut readiness_pipe: File) -> Result<()> {
+    let (non_blocking_writer, _tracing_worker_guard) =
+        tracing_appender::non_blocking(fs::File::create(config.log_file_path)?);
+    tracing_subscriber::fmt()
+        .with_max_level(config.log_level)
+        .with_writer(non_blocking_writer)
+        .with_ansi(false)
+        .with_line_number(true)
+        .with_level(true)
+        .init();
+
+    let socket = DaemonSocket::new(&TeliodDaemon::get_ipc_socket_path()?)?;
+
+    let _telio = Device::new(
+        Features::default(),
+        // TODO: replace this with some real event handling
+        move |event| info!("Incoming event: {:?}", event),
+        None,
+    );
+
+    info!("Entered event loop");
+
+    if let Err(err) = readiness_pipe.write(DAEMON_START_MSG.as_bytes()) {
+        warn!("Could not send \"ready\" message to API process: {:?}", err)
+    }
+
+    // This closes the daemon's pipe end
+    drop(readiness_pipe);
+
+    loop {
+        let mut connection = match socket.accept() {
+            Ok(conn) => conn,
+            Err(err) => {
+                error!("Failed to accept connection: {:?} - stopping daemon", err);
+                break;
+            }
+        };
+
+        let command_str = connection.read_command()?;
+
+        if let Ok(command) = serde_json::from_str::<Cmd>(&command_str) {
+            match command {
+                Cmd::Start { .. } => {
+                    warn!("Start command should never be passed to the daemon");
+                    connection.respond("ERR".to_owned())?;
+                }
+                Cmd::HelloWorld { name } => {
+                    info!("Hello {}!", name);
+                    connection.respond("OK".to_owned())?;
+                }
+                Cmd::Stop => {
```

Comment:
IMHO, it would be nice for the daemon to react to `SIGTERM` instead of specific command, as this is standard method for signaling for process, that it needs to shutdown by standard managers like dumb-init/procd/systemd etc.

Reply:
Hmm, maybe both?

Reply:
for `teliod` specifically it would be rather strange to have stop. As under normal circumstances - `teliod stop` would cause the process to stop, which later would be treated as daemon crash by process manager (dumb-init/systemd/procd) which, under normal circumstances, would restart it. Thus `teliod stop` command would be more like `teliod restart` :D which I guess could be a bit confusing.